Tasks/Common/ServiceFabricHelpers/Connect-ServiceFabricClusterFromServiceEndpoint.ps1

```diff
@@ -0,0 +1,182 @@
+        {
+            $certPassword = $ConnectedServiceEndpoint.Auth.Parameters.CertificatePassword
+        }
+
+        # Explicitly set the key storage to use UserKeySet.  This will ensure the private key is stored in a folder location which the user has access to.
+        # If we don't explicitly set it to UserKeySet, it's possible the MachineKeySet will be used which the user doesn't have access to that folder location, resulting in an access denied error.
+        $certificate.Import($bytes, $certPassword, [System.Security.Cryptography.X509Certificates.X509KeyStorageFlags]::UserKeySet)
+    }
+    catch
+    {
+        throw (Get-VstsLocString -Key ErrorOnCertificateImport -ArgumentList $_)
+    }
+
+    # Add the certificate to the cert store.
+    $store = New-Object System.Security.Cryptography.X509Certificates.X509Store($storeName, $storeLocation)
+    $store.Open(([System.Security.Cryptography.X509Certificates.OpenFlags]::ReadWrite))
+    try
+    {
+        $store.Add($certificate)
+    }
+    finally
+    {
+        $store.Close()
+        $store.Dispose()
+    }
+
+    Write-Host (Get-VstsLocString -Key ImportedCertificate -ArgumentList $certificate.Thumbprint)
+
+    # Override the certificate-related cluster connection parameters to known and supported values
+    $clusterConnectionParameters["FindType"] = "FindByThumbprint"
+    $clusterConnectionParameters["FindValue"] = $certificate.Thumbprint
+    $clusterConnectionParameters["StoreName"] = $storeName.ToString()
+    $clusterConnectionParameters["StoreLocation"] = $storeLocation.ToString()
+
+    return $certificate
+}
+
+function Connect-ServiceFabricClusterFromServiceEndpoint {
+    [CmdletBinding()]
+    param(
+        [Hashtable]
+        $ClusterConnectionParameters,
+
+        $ConnectedServiceEndpoint
+    )
+
+    Trace-VstsEnteringInvocation $MyInvocation
+    try {
+
+        $regKey = "HKLM:\SOFTWARE\Microsoft\Service Fabric SDK"
+        if (!(Test-Path $regKey))
+        {
+            throw (Get-VstsLocString -Key ServiceFabricSDKNotInstalled)
+        }
+
+        $connectionEndpointUrl = [System.Uri]$ConnectedServiceEndpoint.Url
+        # Override the publish profile's connection endpoint with the one defined on the associated service endpoint
+        $clusterConnectionParameters["ConnectionEndpoint"] = $connectionEndpointUrl.Authority # Authority includes just the hostname and port
+
+        # Configure cluster connection pre-reqs
+        if ($ConnectedServiceEndpoint.Auth.Scheme -ne "None")
+        {
+            # Add server cert thumbprint (common to both auth-types)
+            if ($ConnectedServiceEndpoint.Auth.Parameters.ServerCertThumbprint)
+            {
+                $clusterConnectionParameters["ServerCertThumbprint"] = $ConnectedServiceEndpoint.Auth.Parameters.ServerCertThumbprint
+            }
+
+            # Add auth-specific parameters
+            if ($ConnectedServiceEndpoint.Auth.Scheme -eq "UserNamePassword")
```

Comment:
Perhaps, but this is existing code I just moved. I think it might break existing definitions to try and change that.

---
In reply to: [99753938](https://github.com/Microsoft/vsts-tasks/pull/3527#discussion_r99753938) [](ancestors = 99753938)